src/ucs/config/parser.c

```diff
@@ -1076,33 +1076,56 @@ static size_t ucs_config_key_value_count(const ucs_config_key_field_t *cfg_key)
     return count;
 }
 
+static ucs_status_t
+ucs_config_key_value_set_common(const ucs_config_key_value_param_t *param,
+                                void *dest, uint8_t *key_set, const char *value)
+{
+    const ucs_config_key_field_t *cfg_key = param->keys;
+    size_t index;
+
+    for (index = 0; cfg_key->name; ++cfg_key, ++index) {
+        if (!key_set[index]) {
+            if (param->parser.read(value,
+                                UCS_PTR_BYTE_OFFSET(dest, cfg_key->offset),
+                                param->parser.arg) != 1) {
+                return UCS_ERR_INVALID_PARAM;
+            }
+            key_set[index] = 1;
+        }
+    }
+    return UCS_OK;
+}
+
 /**
  * Given key-value pair, find appropriate destination for storing value in the
  * temporary values buffer, and save it there
  */
 static ucs_status_t
-ucs_config_key_value_set(const ucs_config_key_field_t *cfg_key,
-                         const char **values, size_t key_count,
-                         const char *key, const char *value)
+ucs_config_key_value_set(const ucs_config_key_value_param_t *param, void *dest,
+                         uint8_t *key_set, const char *key, const char *value)
 {
+    const ucs_config_key_field_t *cfg_key = param->keys;
     size_t index;
 
-    if (NULL == value) {
-        /* Empty value means that key is a generic value for all keys */
-        for (index = 0; index < key_count; ++index) {
-            values[index] = key;
-        }
-        return UCS_OK;
-    } else {
-        /* Key configured, save value for specific key */
-        for (index = 0; index < key_count; ++index, ++cfg_key) {
-            if (!strcmp(cfg_key->name, key)) {
-                values[index] = value;
-                return UCS_OK;
+    for (index = 0; cfg_key->name; ++cfg_key, ++index) {
+        if (!strcmp(cfg_key->name, key)) {
+            /* Multiple values per key are not allowed */
+            if (key_set[index]) {
+                ucs_error("Key '%s' is configured more than once", key);
+                return UCS_ERR_INVALID_PARAM;
             }
+
+            if (param->parser.read(value,
+                                   UCS_PTR_BYTE_OFFSET(dest, cfg_key->offset),
+                                   param->parser.arg) != 1) {
+                return UCS_ERR_INVALID_PARAM;
+            }
+
+            key_set[index] = 1;
```

Comment:
IMO better not pass key_set to this function, it became too complicated.
a simple function can search for a key by name and return its index or -1 if not found

Reply:
Yes, it's doable of course, but then the `ucs_config_sscanf_key_value` function does everything and becomes quite lengthy.

Initially I liked the idea of using flags, but at the end IMO it makes the code longer, more involved and harder to reason about. So after implementing this proposal, I think the original approach was actually cleaner:
read in a single place, no gotos.
Just compare with what we have now:
```
    tokens = ucs_string_split(tokens, TOKEN_DELIM, 1, &token);
    while (NULL != token) {
        ucs_string_split(token, VALUE_DELIM, 2, &key, &value);

        /* Remember last value in the values array */
        if (UCS_OK != ucs_config_key_value_set(param->keys, values, key_count,
                                               key, value)) {
            return 0;
        }
        tokens = ucs_string_split(tokens, TOKEN_DELIM, 1, &token);
    }

    /* Verify that there is a value for each key. Do it before actual set, so
     * that we don't need to revert in case of failure */
    for (index = 0; index < key_count; ++index) {
        if (NULL == values[index]) {
            ucs_error("No value configured for key '%s'",
                      param->keys[index].name);
            return 0;
        }
    }

    /* Now apply stored temporary values, exactly once per each key */
    for (index = 0; index < key_count; ++index) {
        if (param->parser.read(values[index],
                               (char *)dest + param->keys[index].offset,
                               param->parser.arg) != 1) {
            /* Revert already made settings */
            while (index-- > 0) {
                param->parser.release((char *)dest + param->keys[index].offset,
                                      param->parser.arg);
            }
            return 0;
        }
    }
    return 1;
```

Reply:
Ok, I have just pushed the updated code that addresses your comments.
But personally I would prefer to rollback to the version without flags - it's just not worth it

Reply:
I agree the above code (wo the flags looks cleaner),
1. maybe remove ucs_config_key_value_set helper function?
2. seems we still need to revert if parsing failed, so maybe can unite the check that each key has a value with the parsing?

Reply:
Ok, thanks, I will revert
1. ok
2. will try